File: app/src/main/java/com/project/school/buildingexample/MainActivity.java
package com.project.school.buildingexample;

import android.os.Bundle;
import android.util.Log;

import com.jme3.math.Quaternion;

import eu.kudan.kudan.ARAPIKey;
import eu.kudan.kudan.ARActivity;
import eu.kudan.kudan.ARArbiTrack;
import eu.kudan.kudan.ARArbiTrackListener;
import eu.kudan.kudan.ARGyroPlaceManager;
import eu.kudan.kudan.ARImageNode;
import eu.kudan.kudan.ARImageTrackable;
import eu.kudan.kudan.ARImageTrackableListener;
import eu.kudan.kudan.ARImageTracker;
import eu.kudan.kudan.ARLightMaterial;
import eu.kudan.kudan.ARMeshNode;
import eu.kudan.kudan.ARModelImporter;
import eu.kudan.kudan.ARModelNode;
import eu.kudan.kudan.ARTexture2D;

public class MainActivity extends ARActivity implements ARArbiTrackListener {

    boolean firstRun = true;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ARAPIKey key = ARAPIKey.getInstance();
        key.setAPIKey("LLTYKN8CSOhlIP9lXuXgBkICJ9GMQWE2j0lYaEC4HOmkACBSQYkLdmMLcbe+RP2hAYuHvlbwXZGnOhOWYnBtiTPDLLaCbIG01ZQUbecVHSFHS2TEkaU+9X9RFZPMu/Q8i4jc3icSq3xTHY3+Ux+ZK8FlKHelDV9yl9CZAbo63PdXiCIGIrtj5UTq7eNvrVht/ZPSy6m/Dg3l2byzROut9fUzslvSecOL5uwqhyPH0ecZQ8fuUPiogBUbSoak0BeLJhgrjTiKu+nugqbqHBOpArkKymrJAtSh26dkc8hboiICBlQdW5swaVPZ/G4GT6DLxSAWEUrL7giORgo3lZQAkMkNR5ahi1XmkF+vF0ZoPhw8Y4lsrLG1flReWe5BEalovoo9DhQ0XRhWvZzB7RFRfjdJQrlJ77NKcZVBZp16Zo4wxJgou+o+GnC3t61tQ6AvNh8S/4NQ1Mm18GwGqklSz4+Zf/W3bFqzECyHN50ldws4S2ZdGT4SK0dHbIl8iLMMzEGlorhdiyyR1ungpbwe1JSDq4JFrXkpT8E9geZQPeKlAyHT9mQrymm8lbVTiWBZJX5Q1gRvLzMLpnXkA+LFUiyzHuGtdkWj3FkZQ7hVPgpKq29N3Q1orGXo6hFrGDRkBqDNqtnafnwEzNzjdpYLT4hypANZkXVAueQnZmRi8qM=");
        setContentView(R.layout.activity_main);

    }

    @Override
    public void setup()
    {
        super.setup();
        ARImageTrackable trackable = new ARImageTrackable("StarWars");
        trackable.loadFromAsset("target.png");

// Get instance of image tracker manager
        ARImageTracker trackableManager = ARImageTracker.getInstance();
        trackableManager.initialise();

// Add image trackable to image tracker manager
        trackableManager.addTrackable(trackable);

        ARModelImporter modelImporter = new ARModelImporter();
        modelImporter.loadFromAsset("ben.jet");
        ARModelNode modelNode = (ARModelNode)modelImporter.getNode();
        modelNode.rotateByDegrees(-90,0,0,1);
        modelNode.scaleByUniform(0.3f);
        modelNode.setName("BigBen");

// Load model texture
        ARTexture2D texture2D = new ARTexture2D();
        texture2D.loadFromAsset("bigBenTexture.png");

// Apply model texture file to model texture material and add ambient lighting
        ARLightMaterial material = new ARLightMaterial();
        material.setTexture(texture2D);
        material.setAmbient(0.8f, 0.8f, 0.8f);

// Apply texture material to models mesh nodes
        for (ARMeshNode meshNode : modelImporter.getMeshNodes()){
            meshNode.setMaterial(material);
        }

        // Initialise the image node with our image
        ARImageNode imageNode = new ARImageNode("hearth.png");
        imageNode.setName("Cow");
        imageNode.setPosition(1500,0,500);

// Add the image node as a child of the trackable's world
// Add model node to image trackable
        trackable.getWorld().addChild(modelNode);
        modelNode.addChild(imageNode);

        // Initialise ArbiTrack
        ARArbiTrack arbiTrack = ARArbiTrack.getInstance();
//Add the activity as an ArbiTrack delegate
        arbiTrack.initialise();

        arbiTrack.addListener(this);
        arbiTrack.setTargetNode(trackable.getWorld());
        trackable.addListener(new ARImageTrackableListener() {
            @Override
            public void didDetect(ARImageTrackable arImageTrackable) {
                ARArbiTrack arbiTrack = ARArbiTrack.getInstance();
                arbiTrack.start();
            }

            @Override
            public void didTrack(ARImageTrackable arImageTrackable) {

            }

            @Override
            public void didLose(ARImageTrackable arImageTrackable) {

            }
        });
    }

    @Override
    public void arbiTrackStarted() {
            Log.e("Arbi","Arbitrack Started");
            if (firstRun)
            {
                ARImageTrackable legoTrackable = ARImageTracker.getInstance().findTrackable("StarWars");


                ARArbiTrack arbiTrack = ARArbiTrack.getInstance();


                ARModelNode benNode = (ARModelNode) legoTrackable.getWorld().findChildByName("BigBen");


                Quaternion benFullOrientation = benNode.getWorld().getWorldOrientation().mult(benNode.getWorldOrientation());
                benNode.setOrientation(arbiTrack.getWorld().getOrientation().inverse().mult(benFullOrientation));



                benNode.remove();

                arbiTrack.getWorld().addChild(benNode);

                firstRun = false;
            }
    }
}

File: app/src/main/java/com/project/school/buildingexample/MainActivity.kt
package com.project.school.buildingexample

import android.os.Bundle
import android.support.v4.view.GestureDetectorCompat
import android.view.GestureDetector
import android.view.MotionEvent
import eu.kudan.kudan.*
import io.reactivex.Observable
import java.util.concurrent.TimeUnit

/*
TODO
    Zistit ako funguje GPS tracker v kudane, hlavne ako presne a či vôbec


 */
class MainActivity : ARActivity(), ARArbiTrackListener {

    private val intervalObservable = Observable.interval(6, TimeUnit.MILLISECONDS)
    private val gestureDetector by lazy {
        GestureDetectorCompat(this,object : GestureDetector.OnGestureListener {
            override fun onShowPress(p0: MotionEvent?) {
            }

            override fun onSingleTapUp(p0: MotionEvent?): Boolean {
                isRaining = !isRaining
                val benNode = ARArbiTrack.getInstance().world.findChildByName(ARObjects.BigBen.name)
                return true
            }

            override fun onDown(p0: MotionEvent?): Boolean {
                return true
            }

            override fun onFling(p0: MotionEvent?, p1: MotionEvent?, p2: Float, p3: Float): Boolean {
                return true
            }

            override fun onScroll(p0: MotionEvent?, p1: MotionEvent?, p2: Float, p3: Float): Boolean {
                val ben = ARArbiTrack.getInstance().world.findChildByName(ARObjects.BigBen.name).apply {
                    rotateByDegrees(p2-p3, 0f, 1f, 0f)
                }
                return true

            }

            override fun onLongPress(p0: MotionEvent?) {
            }
        });
    }
    private val trackableId = "Tracker"
    private var isRaining = false
    private val trackable by lazy {
        ARImageTrackable(trackableId).apply {
            loadFromAsset("target.png")
        }
    }

    private var firstRun = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        gestureDetector.onTouchEvent(event)
        return super.onTouchEvent(event)
    }
    override fun setup() {
        super.setup()
        with(ARImageTracker.getInstance()) {
            initialise()
            addTrackable(trackable)
        }

        val bigBen = createBigBen()
        val hearts = listOf(
            createHearth(1500f, 9000f, 500f),
            createHearth(-1500f, 6500f, 750f),
            createHearth(2000f, 1500f, -500f),
            createHearth(1250f, 3400f, -750f),
            createHearth(-1250f, 350f, 400f),
            createHearth(-2000f, 4500f, -500f),
            createHearth(-500f, 2500f, -750f),
            createHearth(1250f, 5350f, 400f),
            createHearth(1250f, 4600f, -400f),
            createHearth(-1200f, 7250f, 1500f)
        )
        val cloud = createCloud(0f,10000f,0f);

        bigBen.addChild(cloud)
        hearts.forEach(bigBen::addChild)

        trackable.world.addChildren(listOf(bigBen))
        ARGyroPlaceManager.getInstance().initialise()

        with(ARArbiTrack.getInstance()) {
            initialise()
            addListener(this@MainActivity)
            targetNode = trackable.world
        }

        trackable.addListener(arImageListener)
    }

    override fun arbiTrackStarted() {
        if (firstRun) {

            val trackable = ARImageTracker.getInstance().findTrackable(trackableId)
            val arbiTrack = ARArbiTrack.getInstance()

            val cloud = trackable.world.findChildByName(ARObjects.Cloud.name)
            val benNode = trackable.world.findChildByName(ARObjects.BigBen.name)

            val benOrientation = benNode.world.worldOrientation.mult(benNode.worldOrientation)
            benNode.orientation = arbiTrack.world.orientation.inverse().mult(benOrientation)

            arbiTrack.world.addChildren(listOf(benNode))

            intervalObservable.subscribe {
                benNode
                    .children
                    .filter { it.isHearth() }
                    .forEach { it.rainAnimation() }
            }

            firstRun = false
        }

    }

    private val arImageListener = object : ARImageTrackableListener {
        override fun didDetect(arImageTrackable: ARImageTrackable) {
            val arbiTrack = ARArbiTrack.getInstance()
            arbiTrack.start()
        }

        override fun didTrack(arImageTrackable: ARImageTrackable) {

        }

        override fun didLose(arImageTrackable: ARImageTrackable) {

        }

    }

    private fun createCloud(x: Float, y: Float, z: Float): ARModelNode {
        val modelImporter = ARModelImporter().apply {
            loadFromAsset("cloud.jet")

        }

        val texture2D = ARTexture2D()
        texture2D.loadFromAsset("clouds.png")

        val material = ARLightMaterial().apply {
            setAmbient(0.8f, 0.8f, 0.8f)
            setTexture(texture2D)
        }

        // Apply texture material to models mesh nodes
        modelImporter.meshNodes.forEach {
            it.material = material
        }

        return modelImporter.node.apply {
            name = ARObjects.Cloud.name
            setPosition(x, y, z)
            scaleByUniform(.5f)
        }

    }

    private fun createBigBen(): ARModelNode {
        val modelImporter = ARModelImporter()
        modelImporter.loadFromAsset("ben.jet")

        val modelNode = modelImporter.node.apply {
            rotateByDegrees(-90f, 0f, 0f, 0f)
            scaleByUniform(0.1f)
            setOrientation(1f, 1f, 1f, 1f)
            rotateByDegrees(-90f, 0f, 0f, 1f)
            name = ARObjects.BigBen.name
        }

        // Load model texture
        val texture2D = ARTexture2D()
        texture2D.loadFromAsset("bigBenTexture.png")

        // Apply model texture file to model texture material and add ambient lighting
        val material = ARLightMaterial().apply {
            setTexture(texture2D)
            setAmbient(0.8f, 0.8f, 0.8f)
        }

        // Apply texture material to models mesh nodes
        modelImporter.meshNodes.forEach {
            it.material = material
        }

        return modelNode
    }

    private fun createHearth(x: Float, y: Float, z: Float) = ARImageNode("heart.png").apply {
        name = ARObjects.Hearth.name
        setPosition(x, y, z)
        scaleByUniform(0.4f)
        visible = isRaining
    }

    private fun ARNode.isHearth() = name.contentEquals(ARObjects.Hearth.name)

    private fun ARNode.rainAnimation(incerment: Int = 10) {
        val x = position.getX()
        val y = position.getY()
        val z = position.getZ()
        if (y < 0){
            setPosition(x, 10000f, z)
            visible = isRaining
        }else
            setPosition(x, y - incerment, z)
    }

}

enum class ARObjects {
    Hearth, BigBen, Cloud
}